README.md
# Bloom

### A highly efficient bloom filter implementation for Go

[![GoDoc](https://godoc.org/github.com/DCSO/bloom?status.svg)](http://godoc.org/github.com/DCSO/bloom)
![Build Status](https://github.com/DCSO/bloom/actions/workflows/go.yml/badge.svg)

Bloom is a simple tool that provides a very efficient implementation of Bloom filters for the go language.
It provides a command line tool that can be used to easily create Bloom filters with desired capacity
and false positive probability. Values can be added to filters through standard input, which makes it
easy to use the tool in a pipeline workflow.

# Usage

    NAME:
       Bloom Filter - Utility to work with bloom filters

    USAGE:
       bloom [global options] command [command options] [arguments...]

    VERSION:
       0.2.2

    COMMANDS:
         create, cr         Create a new Bloom filter and store it in the given filename.
         insert, i          Inserts new values into an existing Bloom filter.
         join, j, merge, m  Joins two Bloom filters into one.
         check, c           Checks values against an existing Bloom filter.
         set-data, sd       Sets the data associated with the Bloom filter.
         get-data, gd       Prints the data associated with the Bloom filter.
         show, s            Shows various details about a given Bloom filter.
         help, h            Shows a list of commands or help for one command

    GLOBAL OPTIONS:
       --gzip, --gz                      compress bloom file with gzip
       --interactive, -i                 interactively add values to the filter
       --split, -s                       split the input string
       --each, -e                        print each match of a split string individually
       --delimiter value, -d value       delimiter to use for splitting (default: ",")
       --fields value, -f value          fields of split output to use in filter (a single number or a comma-separated list of numbers, zero-indexed)
       --print-fields value, --pf value  fields of split output to print for a successful match (a single number or a comma-separated list of numbers, zero-indexed).
       --help, -h                        show help
       --version, -v                     print the version


# Examples

To create a new bloom filter with a desired capacity and false positive probability, you can use the `create` command:

    #will create a gzipped Bloom filter with 100.000 capacity and a 0.1 % false positive probability
    bloom --gzip create -p 0.001 -n 100000 test.bloom.gz

To insert values, you can use the `insert` command and pipe some input to it (each line will be treated as one value):

    cat values | bloom --gzip insert test.bloom.gz

You can also interactively add values to the filter by specifying the `--interactive` command line option:

    bloom --gzip --interactive insert test.bloom.gz

To check if a given value or a list of values is in the filter, you can use the `check` command:

    cat values | bloom --gzip check test.bloom.gz

This will return a list of all values in the filter.

# Advanced Usage

Sometimes it is useful to attach additional information to a string that we want to check against the Bloom filter,
such as a timestamp or the original line content. To make passing along this additional information easier within
a shell context, the Bloom tool provides an option for splitting the input string by a given delimiter and checking
the filter against the resulting field values. Example:

    # will check the Bloom filter for the values foo, bar and baz
    cat "foo,bar,baz" | bloom -s filter.bloom

    # uses a different delimiter (--magic-delimiter--)
    cat "foo--ac5ba--bar--ac5ba--baz" | bloom  -d "--ac5ba--" -s filter.bloom

    # will check the Bloom filter against the second field value only
    cat "foo,bar,baz" | bloom -f 1 -s filter.bloom

    # will check the Bloom filter against the second and third field values only
    cat "foo,bar,baz" | bloom -f 1,2 -s filter.bloom

    # will print one line for each field value that matched against the filter
    cat "foo,bar,baz" | bloom -e -s filter.bloom

    # will print the last field value for each line whose fields matched against the filter
    cat "foo,bar,baz" | bloom -e -s --pf -1 filter.bloom

This functionality is especially handy when using CSV data, as it allows you to filter CSV rows by checking individual
columns against the filter without having to use external tools to split and reassemble the lines.

# Installation

## Installation on Debian-based systems

Debian [command line tool](https://tracker.debian.org/pkg/golang-github-dcso-bloom):

    sudo apt install golang-github-dcso-bloom-cli

## Installation via `go get`:

    go get github.com/DCSO/bloom/...

## Installation from source

These need to be run from within the `GOPATH` source directory for this project (e.g. `$GOPATH/src/github.com/DCSO/bloom`). To install the command line tool:

    make install

To run the tests:

    make test

To run the benchmarks:

    make bench

# Cross-Compiling

To compile a binary, simply specify the target architecture and go:

    #Windows, 64 bit
    env GOOS=windows GOARCH=amd64 go build -v -o bloom.exe github.com/DCSO/bloom
    #Windows, 32 bit
    env GOOS=windows GOARCH=i386 go build -v -o /tmp/bloom github.com/DCSO/bloom


CHANGELOG.md
# Changelog

## v0.2.4 (2021-06-29)

- Add Go module support
- Switch CI/CD to GitHub Actions
- Update documentation

## v0.2.3 (2019-01-10)

- Add test for fingerprint
- Update documentation
- Improve robustness regarding broken input files

## v0.2.2 (2018-10-24)

- Remove dead code

## v0.2.0 (2017-08-14)

- Add 'bloom join' command line tool
- Make it possible to store arbitary data in a Bloom filter (useful for
  associating meta-data with a filter)
- Remove HTTP test
- Use scientific notation for FP probability
- Add a 64-bit value reserved for the version bit and additional flags
- Use 64-bit integers everywhere
- Switched to new, more robust hashing scheme

## v0.1.0 (2017-06-27)

- Initial open source release


io_test.go
// DCSO go bloom filter
// Copyright (c) 2017, DCSO GmbH

package bloom

import (
	"io/ioutil"
	"os"
	"regexp"
	"testing"
)

func checkResults(t *testing.T, bf *BloomFilter) {
	for _, v := range []string{"foo", "bar", "baz"} {
		if !bf.Check([]byte(v)) {
			t.Fatalf("value %s expected in filter but wasn't found", v)
		}
	}
	if bf.Check([]byte("")) {
		t.Fatal("empty value not expected in filter but was found")
	}
	if bf.Check([]byte("12345")) {
		t.Fatal("missing value not expected in filter but was found")
	}
}

func TestFromReaderFile(t *testing.T) {
	f, err := os.Open("testdata/test.bloom")
	if err != nil {
		t.Fatal(err)
	}
	defer f.Close()
	bf, err := LoadFromReader(f, false)
	if err != nil {
		t.Fatal(err)
	}
	checkResults(t, bf)
}

func TestFromReaderCorruptFile(t *testing.T) {
	f, err := os.Open("testdata/broken.bloom")
	if err != nil {
		t.Fatal(err)
	}
	defer f.Close()
	_, err = LoadFromReader(f, false)
	if err == nil {
		t.Fatal("error expected")
	}
	r, _ := regexp.Compile("is too high")
	if !r.MatchString(err.Error()) {
		t.Fatalf("wrong error message: %s", err.Error())
	}
}

func testFromSerialized(t *testing.T, gzip bool) {
	bf := Initialize(100, 0.0001)
	for _, v := range []string{"foo", "bar", "baz"} {
		bf.Add([]byte(v))
	}
	tmpfile, err := ioutil.TempFile("", "test")
	if err != nil {
		t.Fatal(err)
	}
	defer os.Remove(tmpfile.Name())

	err = WriteFilter(&bf, tmpfile.Name(), gzip)
	if err != nil {
		t.Fatal(err)
	}

	loadedBf, err := LoadFilter(tmpfile.Name(), gzip)
	if err != nil {
		t.Fatal(err)
	}
	checkResults(t, loadedBf)
}

func TestFromSerialized(t *testing.T) {
	testFromSerialized(t, false)
}

func TestFromSerializedZip(t *testing.T) {
	testFromSerialized(t, true)
}

func TestFromReaderFileZip(t *testing.T) {
	f, err := os.Open("testdata/test.bloom.gz")
	if err != nil {
		t.Fatal(err)
	}
	defer f.Close()
	bf, err := LoadFromReader(f, true)
	if err != nil {
		t.Fatal(err)
	}
	checkResults(t, bf)
}

func TestFromBytes(t *testing.T) {
	testBytes, err := ioutil.ReadFile("testdata/test.bloom")
	if err != nil {
		t.Fatal(err)
	}
	bf, err := LoadFromBytes(testBytes, false)
	if err != nil {
		t.Fatal(err)
	}
	checkResults(t, bf)
}

func TestFromFile(t *testing.T) {
	bf, err := LoadFilter("testdata/test.bloom", false)
	if err != nil {
		t.Fatal(err)
	}
	checkResults(t, bf)
}


bloom_test.go
// DCSO go bloom filter
// Copyright (c) 2017, DCSO GmbH

package bloom

import (
	"bytes"
	"io/ioutil"
	"log"
	"math"
	"math/rand"
	"os"
	"path/filepath"
	"strings"
	"testing"
)

func TestFingerprinting(t *testing.T) {
	filter := Initialize(100000, 0.01)
	fp := make([]uint64, 7)
	expected := [7]uint64{20311, 36825, 412501, 835777, 658914, 853361, 307361}
	filter.Fingerprint([]byte("bar"), fp)
	for i, v := range fp {
		if v != expected[i] {
			t.Errorf("Wrong fingerprint: %d vs. %d", v, expected[i])
			break
		}
	}
}

func TestInitialization(t *testing.T) {
	filter := Initialize(10000, 0.001)
	if filter.k != 10 {
		t.Error("k does not match expectation!")
	}
	if filter.m != 143775 {
		t.Error("m does not match expectation: ", filter.m)
	}
	if filter.M != uint64(math.Ceil(float64(filter.m)/64)) {
		t.Error("M does not match expectation: ", filter.M)
	}
	for i := uint64(0); i < filter.M; i++ {
		if filter.v[i] != 0 {
			t.Error("Filter value is not initialized to zero!")
		}
	}
}

func checkFilters(a BloomFilter, b BloomFilter, t *testing.T) bool {
	if b.n != a.n ||
		b.p != a.p ||
		b.k != a.k ||
		b.m != a.m ||
		b.M != a.M ||
		!bytes.Equal(b.Data, a.Data) {
		return false
	}
	for i := uint64(0); i < a.M; i++ {
		if a.v[i] != b.v[i] {
			return false
		}
	}
	return true
}

func serializeToBuffer(filter BloomFilter) (*BloomFilter, error) {
	var buf bytes.Buffer
	filter.Write(&buf)
	var newFilter BloomFilter
	newFilter.Read(&buf)
	return &newFilter, nil
}

func serializeToDisk(filter BloomFilter) (*BloomFilter, error) {
	tempFile, err := ioutil.TempFile("", "filter")
	if err != nil {
		return nil, err
	}
	defer os.Remove(tempFile.Name())
	filter.Write(tempFile)
	tempFile.Sync()
	tempFile.Seek(0, 0)
	var newFilter BloomFilter
	err = newFilter.Read(tempFile)
	if err != nil {
		return nil, err
	}
	return &newFilter, nil
}

func TestSerialization(t *testing.T) {
	capacity := uint64(100000)
	p := float64(0.01)
	samples := uint64(1000)
	filter, _ := GenerateExampleFilter(capacity, p, samples)

	newFilter, err := serializeToBuffer(filter)
	if err != nil {
		t.Error("Cannot serialize filter to buffer!")
		return
	}

	if !checkFilters(filter, *newFilter, t) {
		t.Error("Filters do not match!")
	}

	newFilter, err = serializeToDisk(filter)

	if err != nil {
		t.Error("Cannot serialize filter to file!")
		return
	}

	if !checkFilters(filter, *newFilter, t) {
		t.Error("Filters do not match!")
	}

	filter.Add(GenerateTestValue(100))
	newFilter.Add(GenerateTestValue(100))
	newFilter, err = serializeToDisk(filter)
	if err != nil {
		t.Error("Cannot serialize filter to disk!")
		return
	}

	if !checkFilters(filter, *newFilter, t) {
		t.Error("Filters do not match!")
	}

	filter.Add(GenerateTestValue(100))
	newFilter.Add(GenerateTestValue(100))
	newFilter, err = serializeToDisk(filter)
	if err != nil {
		t.Error("Cannot serialize filter to disk!")
		return
	}

	if !checkFilters(filter, *newFilter, t) {
		t.Error("Filters do not match!")
	}

	checkFilters(filter, *newFilter, t)
}

func TestSerializationToDisk(t *testing.T) {
	capacity := uint64(100000)
	p := float64(0.001)
	samples := uint64(1000)
	filter, _ := GenerateExampleFilter(capacity, p, samples)

	var buf bytes.Buffer

	filter.Write(&buf)

	var newFilter BloomFilter

	newFilter.Read(&buf)

	checkFilters(filter, newFilter, t)
}

func TestSerializationWriteFail(t *testing.T) {
	capacity := uint64(100000)
	p := float64(0.001)
	samples := uint64(1000)
	filter, _ := GenerateExampleFilter(capacity, p, samples)

	dir, err := ioutil.TempDir("", "bloomtest")
	if err != nil {
		log.Fatal(err)
	}
	defer os.RemoveAll(dir)

	tmpfn := filepath.Join(dir, "tmpfile")
	tmpfile, err := os.OpenFile(tmpfn, os.O_CREATE|os.O_RDONLY, 0000)
	if err != nil {
		t.Fatal(err)
	}
	defer tmpfile.Close()

	err = filter.Write(tmpfile)
	if err == nil {
		t.Error("writing to read-only file should fail")
	}
}

func TestSerializationReadFail(t *testing.T) {
	var newFilter BloomFilter

	dir, err := ioutil.TempDir("", "bloomtest")
	if err != nil {
		log.Fatal(err)
	}
	defer os.RemoveAll(dir)

	tmpfn := filepath.Join(dir, "tmpfile")
	tmpfile, err := os.OpenFile(tmpfn, os.O_CREATE, 0777)
	if err != nil {
		t.Fatal(err)
	}
	defer tmpfile.Close()

	err = newFilter.Read(tmpfile)
	if err == nil {
		t.Error("reading from empty file should fail")
	}
}

func GenerateTestValue(length uint64) []byte {
	value := make([]byte, length)
	for i := uint64(0); i < length; i++ {
		value[i] = byte(rand.Int() % 256)
	}
	return value
}

func GenerateExampleFilter(capacity uint64, p float64, samples uint64) (BloomFilter, [][]byte) {
	filter := Initialize(capacity, p)
	filter.Data = []byte("foobar")
	testValues := make([][]byte, 0, samples)
	for i := uint64(0); i < samples; i++ {
		testValue := GenerateTestValue(100)
		testValues = append(testValues, testValue)
		filter.Add(testValue)
	}
	return filter, testValues
}

func GenerateDisjointExampleFilter(capacity uint64, p float64, samples uint64, other BloomFilter) (BloomFilter, [][]byte) {
	filter := Initialize(capacity, p)
	testValues := make([][]byte, 0, samples)
	for i := uint64(0); i < samples; {
		testValue := GenerateTestValue(100)
		if !other.Check(testValue) {
			testValues = append(testValues, testValue)
			filter.Add(testValue)
			i++
		}
	}
	return filter, testValues
}

//This tests the checking of values against a given filter
func TestChecking(t *testing.T) {
	capacity := uint64(100000)
	p := float64(0.001)
	samples := uint64(100000)
	filter, testValues := GenerateExampleFilter(capacity, p, samples)
	fingerprint := make([]uint64, filter.k)
	for _, value := range testValues {
		filter.Fingerprint(value, fingerprint)
		if !filter.CheckFingerprint(fingerprint) {
			t.Error("Did not find test value in filter!")
		}
	}
}

//This tests the checking of values against a given filter after resetting it
func TestReset(t *testing.T) {
	capacity := uint64(100000)
	p := float64(0.001)
	samples := uint64(100000)
	filter, testValues := GenerateExampleFilter(capacity, p, samples)
	filter.Reset()
	fingerprint := make([]uint64, filter.k)
	for _, value := range testValues {
		filter.Fingerprint(value, fingerprint)
		if filter.CheckFingerprint(fingerprint) {
			t.Error("Did not find test value in filter!")
		}
	}
}

//This tests the checking of values against a given filter
//see https://en.wikipedia.org/wiki/Bloom_filter#Probability_of_false_positives
func TestFalsePositives(t *testing.T) {
	capacity := uint64(10000)
	p := float64(0.001)
	fillingFactor := 0.9
	N := uint64(float64(capacity) * fillingFactor)
	filter, _ := GenerateExampleFilter(capacity, p, N)
	pAcceptable := math.Pow(1-math.Exp(-float64(filter.k)*float64(N)/float64(filter.m)), float64(filter.k))
	fingerprint := make([]uint64, filter.k)
	cnt := 0.0
	matches := 0.0
	for {
		cnt++
		value := GenerateTestValue(100)
		filter.Fingerprint(value, fingerprint)
		if filter.CheckFingerprint(fingerprint) {
			matches++
		}
		if cnt > float64(capacity)*10 {
			break
		}
	}
	//this might still fail sometimes...
	//we allow for a probability that is two times higher than the normally acceptable probability
	if matches/cnt > pAcceptable*2 {
		t.Error("False positive probability is too high at ", matches/cnt*100, "% vs ", pAcceptable*100, "%")
	}
}

func TestJoiningRegularMisdimensioned(t *testing.T) {
	a := Initialize(100000, 0.0001)
	b := Initialize(10000, 0.0001)
	err := a.Join(&b)
	if err == nil {
		t.Error("joining filters with different capacity should fail")
	}
	if !strings.Contains(err.Error(), "different dimensions") {
		t.Error("wrong error message returned")
	}
	a = Initialize(100000, 0.0001)
	b = Initialize(100000, 0.001)
	err = a.Join(&b)
	if err == nil {
		t.Error("joining filters with different FP prob should fail")
	}
	if !strings.Contains(err.Error(), "different dimensions") {
		t.Error("wrong error message returned")
	}
	a = Initialize(100000, 0.0001)
	b = Initialize(100000, 0.0001)
	b.k = 1
	err = a.Join(&b)
	if err == nil {
		t.Error("joining filters with different number of hash funcs should fail")
	}
	if !strings.Contains(err.Error(), "different dimensions") {
		t.Error("wrong error message returned")
	}
	a = Initialize(100000, 0.0001)
	b = Initialize(100000, 0.0001)
	b.m = 1
	err = a.Join(&b)
	if err == nil {
		t.Error("joining filters with different number of bits should fail")
	}
	if !strings.Contains(err.Error(), "different dimensions") {
		t.Error("wrong error message returned")
	}
	a = Initialize(100000, 0.0001)
	b = Initialize(100000, 0.0001)
	b.M = 1
	err = a.Join(&b)
	if err == nil {
		t.Error("joining filters with different int array size should fail")
	}
	if !strings.Contains(err.Error(), "different dimensions") {
		t.Error("wrong error message returned")
	}
}

func TestAccessors(t *testing.T) {
	a, _ := GenerateExampleFilter(100000, 0.0001, 10000)
	if a.MaxNumElements() != 100000 {
		t.Error("unexpected capacity in filter")
	}
	if a.NumBits() != 1917011 {
		t.Error("unexpected number of bits in filter")
	}
	if a.NumHashFuncs() != 14 {
		t.Error("unexpected number of hash funcs in filter")
	}
	if a.FalsePositiveProb() != 0.0001 {
		t.Error("unexpected FP prob in filter")
	}
}

func TestJoiningRegular(t *testing.T) {
	a, aval := GenerateExampleFilter(100000, 0.0001, 10000)
	b, bval := GenerateDisjointExampleFilter(100000, 0.0001, 20000, a)
	for _, v := range bval {
		if a.Check(v) {
			t.Errorf("value not missing in joined filter: %s", string(v))
		}
	}
	if a.N != 10000 {
		t.Error("unexpected number of elements in filter")
	}
	if b.N != 20000 {
		t.Error("unexpected number of elements in filter")
	}
	err := a.Join(&b)
	if a.N != 30000 {
		t.Errorf("unexpected number of elements in filter")
	}
	if err != nil {
		t.Fatal(err)
	}
	for _, v := range aval {
		if !a.Check(v) {
			t.Errorf("value not found in joined filter: %s", string(v))
		}
	}
	for _, v := range bval {
		if !a.Check(v) {
			t.Errorf("value not found in joined filter: %s", string(v))
		}
	}
}

//This benchmarks the checking of values against a given filter
func BenchmarkChecking(b *testing.B) {
	capacity := uint64(1e9)
	p := float64(0.001)
	samples := uint64(100000)
	filter, testValues := GenerateExampleFilter(capacity, p, samples)
	fingerprint := make([]uint64, filter.k)
	b.ResetTimer()
	for i := 0; i < b.N; i++ {
		value := testValues[rand.Int()%len(testValues)]
		filter.Fingerprint(value, fingerprint)
		if !filter.CheckFingerprint(fingerprint) {
			b.Error("Did not find test value in filter!")
		}
	}
}

//This benchmarks the checking without using a fixed fingerprint variable (instead a temporary variable is created each time)
func BenchmarkSimpleChecking(b *testing.B) {
	capacity := uint64(1e9)
	p := float64(0.001)
	samples := uint64(100000)
	filter, testValues := GenerateExampleFilter(capacity, p, samples)
	b.ResetTimer()
	for i := 0; i < b.N; i++ {
		value := testValues[rand.Int()%len(testValues)]
		if !filter.Check(value) {
			b.Error("Did not find test value in filter!")
		}
	}
}


bloom/manage.go
// DCSO go bloom filter
// Copyright (c) 2017, DCSO GmbH

package main

import (
	"bufio"
	"bytes"
	"fmt"
	"os"
	"path/filepath"
	"strconv"
	"strings"

	"github.com/DCSO/bloom"
	"gopkg.in/urfave/cli.v1"
)

// BloomParams represents the parameters of the 'bloom' command line tool.
type BloomParams struct {
	gzip           bool
	interactive    bool
	split          bool
	printEachMatch bool
	delimiter      string
	fields         []int
	printFields    []int
}

func exitWithError(message string) {
	fmt.Fprintf(os.Stderr, "Error: %s \n", message)
	os.Exit(-1)
}

func readValuesIntoFilter(filter *bloom.BloomFilter, bloomParams BloomParams) {
	//we determine if the program is run interactively or within a pipe
	stat, _ := os.Stdin.Stat()
	var isTerminal = (stat.Mode() & os.ModeCharDevice) != 0
	//if we are not in an interactive session and this is a terminal, we quit
	if !bloomParams.interactive && isTerminal {
		return
	}
	if bloomParams.interactive {
		fmt.Println("Interactive mode: Enter a blank line [by pressing ENTER] to exit (values will not be stored otherwise).")
	}
	scanner := bufio.NewScanner(os.Stdin)
	for scanner.Scan() {
		line := scanner.Text()
		if line == "" && bloomParams.interactive {
			break
		}
		if bloomParams.split {
			values := strings.Split(line, bloomParams.delimiter)
			for i, value := range values {
				j := i - len(values)

				if len(bloomParams.fields) > 0 {
					if !contains(bloomParams.fields, i) && !contains(bloomParams.fields, j) {
						continue
					}
				}
				filter.Add([]byte(value))
			}
		} else {
			filter.Add([]byte(line))
		}
	}
}

func readInputIntoData(filter *bloom.BloomFilter, bloomParams BloomParams) {
	//we determine if the program is run interactively or within a pipe
	stat, _ := os.Stdin.Stat()
	var isTerminal = (stat.Mode() & os.ModeCharDevice) != 0
	//if we are not in an interactive session and this is a terminal, we quit
	if !bloomParams.interactive && isTerminal {
		return
	}
	if bloomParams.interactive {
		fmt.Println("Interactive mode: Enter a blank line [by pressing ENTER] to exit (values will not be stored otherwise).")
	}
	scanner := bufio.NewScanner(os.Stdin)
	dataBuffer := bytes.NewBuffer([]byte(""))
	for scanner.Scan() {
		line := scanner.Bytes()
		if len(line) == 0 && bloomParams.interactive {
			break
		}
		dataBuffer.Write(line)
		dataBuffer.Write([]byte("\n"))
	}
	filter.Data = dataBuffer.Bytes()
}

func insertIntoFilter(path string, bloomParams BloomParams) {
	filter, err := bloom.LoadFilter(path, bloomParams.gzip)
	if err != nil {
		exitWithError(err.Error())
	}
	readValuesIntoFilter(filter, bloomParams)
	err = bloom.WriteFilter(filter, path, bloomParams.gzip)
	if err != nil {
		exitWithError(err.Error())
	}
}

func updateFilterData(path string, bloomParams BloomParams) {
	filter, err := bloom.LoadFilter(path, bloomParams.gzip)
	if err != nil {
		exitWithError(err.Error())
	}
	readInputIntoData(filter, bloomParams)
	err = bloom.WriteFilter(filter, path, bloomParams.gzip)
	if err != nil {
		exitWithError(err.Error())
	}
}

func getFilterData(path string, bloomParams BloomParams) {
	filter, err := bloom.LoadFilter(path, bloomParams.gzip)
	if err != nil {
		exitWithError(err.Error())
	}
	fmt.Print(string(filter.Data))
}

func contains(s []int, e int) bool {
	for _, a := range s {
		if a == e {
			return true
		}
	}
	return false
}

func checkAgainstFilter(path string, bloomParams BloomParams) {
	filter, err := bloom.LoadFilter(path, bloomParams.gzip)
	if err != nil {
		exitWithError(err.Error())
	}
	scanner := bufio.NewScanner(os.Stdin)
	if bloomParams.interactive {
		fmt.Println("Interactive mode: Enter a blank line [by pressing ENTER] to exit.")
	}
	for scanner.Scan() {
		line := scanner.Text()
		if line == "" && bloomParams.interactive {
			break
		}
		var valuesToCheck []string
		if bloomParams.split {
			valuesToCheck = strings.Split(line, bloomParams.delimiter)
		} else {
			valuesToCheck = make([]string, 1)
			valuesToCheck[0] = line
		}
		printed := false
		prefix := ""
		if bloomParams.interactive {
			prefix = ">"
		}
		for i, value := range valuesToCheck {
			j := i - len(valuesToCheck)
			//we only check fields that are in the "fields" parameters (if defined)
			if len(bloomParams.fields) > 0 {
				if !contains(bloomParams.fields, i) && !contains(bloomParams.fields, j) {
					continue
				}
			}

			if filter.Check([]byte(value)) {
				if bloomParams.printEachMatch {
					fmt.Printf("%s%s\n", prefix, value)
				} else {
					if !printed {
						if len(bloomParams.printFields) > 0 {
							values := make([]string, 0, len(bloomParams.printFields))
							for _, i := range bloomParams.printFields {
								j := i
								if j < 0 {
									j = j + len(valuesToCheck)
								}
								if j >= len(valuesToCheck) || j < 0 {
									continue
								}
								values = append(values, valuesToCheck[j])
							}
							fmt.Printf("%s%s\n", prefix, strings.Join(values, bloomParams.delimiter))
						} else {
							fmt.Printf("%s%s\n", prefix, line)
						}
					}
					printed = true
				}
			}
		}
	}
}

func printStats(path string, bloomParams BloomParams) {
	filter, err := bloom.LoadFilter(path, bloomParams.gzip)
	if err != nil {
		exitWithError(err.Error())
	}
	fmt.Printf("File:\t\t\t%s\n", path)
	fmt.Printf("Capacity:\t\t%d\n", filter.MaxNumElements())
	fmt.Printf("Elements present:\t%d\n", filter.N)
	fmt.Printf("FP probability:\t\t%.2e\n", filter.FalsePositiveProb())
	fmt.Printf("Bits:\t\t\t%d\n", filter.NumBits())
	fmt.Printf("Hash functions:\t\t%d\n", filter.NumHashFuncs())
}

func createFilter(path string, n uint64, p float64, bloomParams BloomParams) {
	filter := bloom.Initialize(n, p)
	readValuesIntoFilter(&filter, bloomParams)
	err := bloom.WriteFilter(&filter, path, bloomParams.gzip)
	if err != nil {
		exitWithError(err.Error())
	}
}

func joinFilters(path string, pathToAdd string, bloomParams BloomParams) {
	filter, err := bloom.LoadFilter(path, bloomParams.gzip)
	if err != nil {
		exitWithError(err.Error())
	}
	filter2, err := bloom.LoadFilter(pathToAdd, bloomParams.gzip)
	if err != nil {
		exitWithError(err.Error())
	}
	err = filter.Join(filter2)
	if err != nil {
		exitWithError(err.Error())
	}
	err = bloom.WriteFilter(filter, path, bloomParams.gzip)
	if err != nil {
		exitWithError(err.Error())
	}
}

func parseFieldIndexes(s string) ([]int, error) {
	fields := strings.Split(s, ",")
	fieldNumbers := make([]int, len(fields))
	for i, field := range fields {
		num, err := strconv.Atoi(field)
		if err != nil {
			return nil, err
		}
		fieldNumbers[i] = num
	}
	return fieldNumbers, nil
}

func parseBloomParams(c *cli.Context) BloomParams {
	var bloomParams BloomParams
	var err error
	bloomParams.gzip = c.GlobalBool("gzip")
	bloomParams.interactive = c.GlobalBool("interactive")
	bloomParams.split = c.GlobalBool("split")
	bloomParams.delimiter = c.GlobalString("delimiter")
	bloomParams.printEachMatch = c.GlobalBool("each")
	if c.GlobalString("fields") != "" {
		bloomParams.fields, err = parseFieldIndexes(c.GlobalString("fields"))
		if err != nil {
			exitWithError(err.Error())
		}
	}
	if c.GlobalString("print-fields") != "" {
		bloomParams.printFields, err = parseFieldIndexes(c.GlobalString("print-fields"))
		if err != nil {
			exitWithError(err.Error())
		}
		//if printFields is set we also set printEachMatch
		if len(bloomParams.printFields) > 0 {
			bloomParams.printEachMatch = false
		}
	}
	return bloomParams
}

func main() {

	app := cli.NewApp()
	app.Name = "Bloom Filter"
	app.Usage = "Utility to work with bloom filters"
	app.Flags = []cli.Flag{
		cli.BoolFlag{
			Name:  "gzip, gz",
			Usage: "compress bloom file with gzip",
		},
		cli.BoolFlag{
			Name:  "interactive, i",
			Usage: "interactively add values to the filter",
		},
		cli.BoolFlag{
			Name:  "split, s",
			Usage: "split the input string",
		},
		cli.BoolFlag{
			Name:  "each, e",
			Usage: "print each match of a split string individually",
		},
		cli.StringFlag{
			Name:  "delimiter, d",
			Value: ",",
			Usage: "delimiter to use for splitting",
		},
		cli.StringFlag{
			Name:  "fields, f",
			Value: "",
			Usage: "fields of split output to use in filter (a single number or a comma-separated list of numbers, zero-indexed)",
		},
		cli.StringFlag{
			Name:  "print-fields, pf",
			Value: "",
			Usage: "fields of split output to print for a successful match (a single number or a comma-separated list of numbers, zero-indexed).",
		},
	}
	app.Commands = []cli.Command{
		{
			Name:    "create",
			Aliases: []string{"cr"},
			Flags: []cli.Flag{
				cli.Float64Flag{Name: "p", Value: 0.01, Usage: "The desired false positive probability."},
				cli.Uint64Flag{Name: "n", Value: 10000, Usage: "The desired capacity."},
			},
			Usage: "Create a new Bloom filter and store it in the given filename.",
			Action: func(c *cli.Context) error {
				path := c.Args().First()
				bloomParams := parseBloomParams(c)
				if path == "" {
					exitWithError("No filename given.")
				}
				path, err := filepath.Abs(path)
				if err != nil {
					return err
				}
				n := c.Uint64("n")
				p := c.Float64("p")
				if n < 0 {
					exitWithError("n cannot be negative.")
				}
				if p < 0 || p > 1 {
					exitWithError("p must be between 0 and 1.")
				}
				createFilter(path, n, p, bloomParams)
				return nil
			},
		},
		{
			Name:    "insert",
			Aliases: []string{"i"},
			Flags:   []cli.Flag{},
			Usage:   "Inserts new values into an existing Bloom filter.",
			Action: func(c *cli.Context) error {
				path := c.Args().First()
				bloomParams := parseBloomParams(c)
				if path == "" {
					exitWithError("No filename given.")
				}
				path, err := filepath.Abs(path)
				if err != nil {
					return err
				}
				insertIntoFilter(path, bloomParams)
				return nil
			},
		},
		{
			Name:    "join",
			Aliases: []string{"j", "merge", "m"},
			Flags:   []cli.Flag{},
			Usage:   "Joins two Bloom filters into one.",
			Action: func(c *cli.Context) error {
				if len(c.Args()) != 2 {
					exitWithError("Two filenames are required.")
				}
				bloomParams := parseBloomParams(c)
				path := c.Args().First()
				if path == "" {
					exitWithError("No first filename given.")
				}
				path, err := filepath.Abs(path)
				if err != nil {
					return err
				}
				pathToAdd := c.Args().Get(1)
				if pathToAdd == "" {
					exitWithError("No second filename given.")
				}
				pathToAdd, err = filepath.Abs(pathToAdd)
				if err != nil {
					return err
				}
				joinFilters(path, pathToAdd, bloomParams)
				return nil
			},
		},
		{
			Name:    "check",
			Aliases: []string{"c"},
			Flags:   []cli.Flag{},
			Usage:   "Checks values against an existing Bloom filter.",
			Action: func(c *cli.Context) error {
				path := c.Args().First()
				bloomParams := parseBloomParams(c)
				if path == "" {
					exitWithError("No filename given.")
				}
				path, err := filepath.Abs(path)
				if err != nil {
					return err
				}
				checkAgainstFilter(path, bloomParams)
				return nil
			},
		},
		{
			Name:    "set-data",
			Aliases: []string{"sd"},
			Flags:   []cli.Flag{},
			Usage:   "Sets the data associated with the Bloom filter.",
			Action: func(c *cli.Context) error {
				path := c.Args().First()
				bloomParams := parseBloomParams(c)
				if path == "" {
					exitWithError("No filename given.")
				}
				path, err := filepath.Abs(path)
				if err != nil {
					return err
				}
				updateFilterData(path, bloomParams)
				return nil
			},
		},
		{
			Name:    "get-data",
			Aliases: []string{"gd"},
			Flags:   []cli.Flag{},
			Usage:   "Prints the data associated with the Bloom filter.",
			Action: func(c *cli.Context) error {
				path := c.Args().First()
				bloomParams := parseBloomParams(c)
				if path == "" {
					exitWithError("No filename given.")
				}
				path, err := filepath.Abs(path)
				if err != nil {
					return err
				}
				getFilterData(path, bloomParams)
				return nil
			},
		},
		{
			Name:    "show",
			Aliases: []string{"s"},
			Flags:   []cli.Flag{},
			Usage:   "Shows various details about a given Bloom filter.",
			Action: func(c *cli.Context) error {
				path := c.Args().First()
				bloomParams := parseBloomParams(c)
				if path == "" {
					exitWithError("No filename given.")
				}
				path, err := filepath.Abs(path)
				if err != nil {
					return err
				}
				printStats(path, bloomParams)
				return nil
			},
		},
	}
	app.Version = "0.2.4"

	app.Run(os.Args)
}
